### ISSUE TITLE
[dist-txns] [failure-testing] unique secondary idx workload: CPU pegged on restart of a killed node

### ISSUE THREAD

**Issue by kmuthukk** [YugaByte/yugabyte-db#485]
Upon restart of a killed node, during distributed unique secondary index workload, the restarted node's CPU becomes pegged 100% and write workload does not recover.

1) This is a on a multi-AZ cluster running in AWS

http://172.151.44.11:7000/tablet-servers

2) It is running a Unique Secondary Index workload with version 1.1.0.2.

```
java -jar yb-sample-apps.jar --workload CassandraUniqueSecondaryIndex --nodes $CIP_ADDR -num_threads_read 1 -num_threads_write 16 --num_unique_keys 100000000 --num_writes 100000000 --nouuid
```

3) I stopped yb-tserver on 172.151.44.11. Workload dipped for a quick moment, but then recovered and kept running... no issues

4) I then restarted yb-tserver on 172.151.44.11. Tablets bootstrapped, leaders balanced, etc. But WRITES are stuck:
```
018-09-28 01:05:23,908 [INFO|com.yugabyte.sample.common.metrics.MetricsTracker|MetricsTracker] Read: 560.57 ops/sec (1.78 ms/op), 629134 total ops  |  Write: 0.00 ops/sec (0.00 ms/op), 3292175 total ops
```

Writes did not recover at all.

CPU utilization on this node was at nearly 98% and stayed that way.

Running  "perf top" highlights these top contenders on this node

```
 18.36%  librocksdb.so                   [.] rocksdb::(anonymous namespace)::BytewiseComparatorImpl::Compare
 17.84%  librocksdb.so                   [.] rocksdb::DBIter::FindNextUserEntryInternal
 14.95%  librocksdb.so                   [.] rocksdb::MergingIterator::Next
 10.74%  libc-2.23.so                    [.] __memcpy_avx_unaligned
  9.98%  librocksdb.so                   [.] rocksdb::BlockIter::ParseNextKey
  5.25%  librocksdb.so                   [.] rocksdb::(anonymous namespace)::TwoLevelIterator::Next
  3.20%  librocksdb.so                   [.] rocksdb::(anonymous namespace)::TwoLevelIterator::SkipEmptyDataBlocksForward
  1.96%  ld.so                           [.] __tls_get_addr
  1.95%  librocksdb.so                   [.] rocksdb::BlockIter::Valid
  1.86%  librocksdb.so                   [.] rocksdb::MergingIterator::Valid
  1.65%  librocksdb.so                   [.] rocksdb::BlockIter::Next
```

A thread dump shows about 25 stacks like: not sure if that's  problem.

```
 0x7fc4e337a0df  (unknown) (/tmp/glibc-20180316-12585-rgaxli/glibc-2.23/signal/../sysdeps/unix/sysv/linux/x86_64/sigaction.c:0)
    @     0x7fc4ed116045  rocksdb::BlockIter::ParseNextKey() (src/yb/rocksdb/table/block.cc:169)
    @     0x7fc4ed149b88  rocksdb::IteratorWrapper::Next() (src/yb/rocksdb/table/iterator_wrapper.h:124)
    @     0x7fc4ed149b88  Next (src/yb/rocksdb/table/two_level_iterator.cc:149)
    @     0x7fc4ed133e45  rocksdb::IteratorWrapper::Next() (src/yb/rocksdb/table/iterator_wrapper.h:124)
    @     0x7fc4ed133e45  rocksdb::MergingIterator::Next() (src/yb/rocksdb/table/merger.cc:201)
    @     0x7fc4ed08ba40  rocksdb::DBIter::FindNextUserEntryInternal(bool) (src/yb/rocksdb/db/db_iter.cc:367)
    @     0x7fc4ed08c44d  rocksdb::DBIter::FindNextUserEntry(bool) (src/yb/rocksdb/db/db_iter.cc:298)
    @     0x7fc4ed08c44d  rocksdb::DBIter::Seek(yb::Slice const&) (src/yb/rocksdb/db/db_iter.cc:757)
    @     0x7fc4ed7792c5  yb::docdb::PerformRocksDBSeek(rocksdb::Iterator*, yb::Slice const&, char const*, int) (src/yb/docdb/docdb_rocksdb_util.cc:267)
    @     0x7fc4ed7da975  yb::docdb::IntentAwareIterator::SeekIntentIterIfNeeded() (src/yb/docdb/intent_aware_iterator.cc:530)
    @     0x7fc4ed7daccc  yb::docdb::IntentAwareIterator::valid() (src/yb/docdb/intent_aware_iterator.cc:546)
    @     0x7fc4ed7c130a  yb::docdb::DocRowwiseIterator::HasNext() const (src/yb/docdb/doc_rowwise_iterator.cc:246)
    @     0x7fc4ed7aa9b8  yb::docdb::QLWriteOperation::ReadColumns(yb::docdb::DocOperationApplyData const&, yb::Schema*, yb::Schema*, yb::QLTableRow*) (src/yb/docdb/doc_operation.cc:2111)
    @     0x7fc4ed7abeaa  yb::docdb::QLWriteOperation::Apply(yb::docdb::DocOperationApplyData const&) (src/yb/docdb/doc_operation.cc:2418)
    @     0x7fc4ed7678ec  yb::docdb::ExecuteDocWriteOperation(std::vector >, std::allocator > > > const&, yb::MonoTime, yb::ReadHybridTime const&, yb::docdb::DocDB const&, yb::docdb::KeyValueWriteBatchPB*, yb::docdb::InitMarkerBehavior, std::atomic*, yb::HybridTime*) (src/yb/docdb/docdb.cc:175)
    @     0x7fc4ee1f7e30  yb::tablet::Tablet::StartDocWriteOperation(yb::tablet::WriteOperation*) (src/yb/tablet/tablet.cc:1588)
    @     0x7fc4ee2034fb  yb::tablet::Tablet::KeyValueBatchFromQLWriteBatch(unique_ptr >) (src/yb/tablet/tablet.cc:888)
    @     0x7fc4ee2044e2  yb::tablet::Tablet::AcquireLocksAndPerformDocOperations(unique_ptr >) (src/yb/tablet/tablet.cc:1158)
```

**Comment by kmuthukk** [comment]
@robertpang 

When the system is doing writes after the restart, the aforementioned yb-tserver logs continue to spew the following:

```
When I run the CUSI (unique sec idx) workload after restart... it is still spewing these errors like crazy:

E0928 06:11:35.523056 23094 process_context.cc:180] SQL Error: Tablet Not Found. Not found: Failed Write(tablet: 1cb7e3157d184fd4b3ca2249888534b0, num_ops: 1, num_attempts: 1) to tablet 1cb7e3157d184fd4b3ca2249888534b0 on tablet server { uuid: 2929187a2a994a6194f0b8484f24dc5d private: [host: "172.151.44.11" port: 9100] cloud_info: placement_cloud: "aws" placement_region: "us-west" placement_zone: "us-west-2b" after 1 attempt(s): Index id 495d9aee5ad94c93aeeb7800093c83f5 not found
INSERT INTO CassandraUniqueSecondaryIndex (k, v) VALUES (?, ?);
      ^^^^
E0928 06:11:35.523073 23116 process_context.cc:180] SQL Error: Tablet Not Found. Not found: Failed Write(tablet: 37b0c2ab95474c9586cf16b2425c781b, num_ops: 1, num_attempts: 1) to tablet 37b0c2ab95474c9586cf16b2425c781b on tablet server { uuid: 2929187a2a994a6194f0b8484f24dc5d private: [host: "172.151.44.11" port: 9100] cloud_info: placement_cloud: "aws" placement_region: "us-west" placement_zone: "us-west-2b" after 1 attempt(s): Index id 495d9aee5ad94c93aeeb7800093c83f5 not found
INSERT INTO CassandraUniqueSecondaryIndex (k, v) VALUES (?, ?);
      ^^^^
E0928 06:11:35.523399 23013 process_context.cc:180] SQL Error: Tablet Not Found. Not found: Failed Write(tablet: 6bf7d37414e1473f845032da864ec4b3, num_ops: 1, num_attempts: 1) to tablet 6bf7d37414e1473f845032da864ec4b3 on tablet server { uuid: 2929187a2a994a6194f0b8484f24dc5d private: [host: "172.151.44.11" port: 9100] cloud_info: placement_cloud: "aws" placement_region: "us-west" placement_zone: "us-west-2b" after 1 attempt(s): Index id 495d9aee5ad94c93aeeb7800093c83f5 not found
```

They are all complaining about the same "Index id 495d9aee5ad94c93aeeb7800093c83f5" which is the "table" id of the index. Is it possible that the logic similar to ybclient that you have for TServer to do index update RPC calls is not able to refresh its cache?

We should do the classic 1-time refresh of the meta-data to account for the case that the tablet of the index table has moved, right?

**Comment by robertpang** [comment]
Fixed in commit [41e009a](https://github.com/YugaByte/yugabyte-db/commit/41e009aaa61f1cdfb0e346cbd99d5562b8be2704)
